activity_calendar/models.py

```diff
@@ -22,45 +30,232 @@ class Meta:
     created_date = models.DateTimeField(auto_now_add=True)
     last_updated_date = models.DateTimeField(auto_now=True)
     
     # The date at which the activity will become visible for all users
-    published_date = models.DateTimeField(default=timezone.now)
+    published_date = models.DateTimeField(default=now_rounded)
 
     # Start and end times
-    start_date = models.DateTimeField(default=timezone.now)
-    end_date = models.DateTimeField(default=timezone.now)
+    start_date = models.DateTimeField(default=now_rounded)
+    end_date = models.DateTimeField(default=later_rounded)
 
     # Recurrence information (e.g. a weekly event)
     # This means we do not need to store (nor create!) recurring activities separately
     recurrences = RecurrenceField(blank=True, default="")
 
     # Maximum number of participants/slots
     # -1 denotes unlimited
-    max_slots = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
+    max_slots = models.IntegerField(default=1, validators=[MinValueValidator(-1)],
         help_text="-1 denotes unlimited slots")
     max_participants = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
         help_text="-1 denotes unlimited participants")
 
-    # Maximum number of slots that someone can create/join
-    max_slots_join_per_participant = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
-        help_text="-1 denotes unlimited slots")
-    max_slots_create_per_participant = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
+    # Maximum number of slots that someone can join
+    max_slots_join_per_participant = models.IntegerField(default=1, validators=[MinValueValidator(-1)],
         help_text="-1 denotes unlimited slots")
     
+    subscriptions_required = models.BooleanField(default=True,
+        help_text="People are only allowed to go to the activity if they register beforehand")
+
+    # auto_create_first_slot = models.BooleanField(default=True,
+    #     help_text="The first slot is automatically created if someone registers for the activity.")
+
+    # Possible slot-creation options:
+    # - Never: Slots can only be created in the admin panel
+    # - Auto: Slots are created automatically. They are only actually created in the DB once a participant joins.
+    #                                          Until that time they do look like real slots though (in the UI)
+    # - Users: Slots can be created by users. Users can be the owner of at most max_slots_join_per_participant slots
+    SLOT_CREATION_OPTIONS = [
+        ("CREATION_NONE",   "Never/By Administrators"),
+        ("CREATION_AUTO",   "Automatically"),
+        ("CREATION_USER",   "By Users"),
+    ]
+
+    # The maximum number of dummy slots that show up if slot_creation=CREATION_AUTO
+    # Dummy slots do not exist in the DB itself, and match their title, description, etc. with the parent activity
```

Comment:
Why do dummy slots not exist in the database?

Reply:
Because that would mean we actually need to create them at some point in time. Requiring admins to create them means they need to create slots for each activity, which was a task we were wanting to avoid.
If the server needs to create them (in the DB), then it needs to be aware of when an activity's occurrence is and actually create the slots some time before that. Creating the slots when an activity is defined is not really an option, as activities can recur indefinitely.

Having the slots show up in the frontend without actually existing, and only creating them if they're being used, seemed like the easier approach.

---
This did make me realise that changing a recurring activity's start-date or recurrence-rules will break all already existing slots for that activity. The former is easy to fix; the latter not so much.

Reply:
I suggest we move a fix for that to a later branch that would already focus on supporting activity editing.

Reply:
Ok, I suggest we move that to the backlog and for now forbid editing recurrences